aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -183,9 +183,104 @@ static void prod_kernel_impl(Tensor& result, const Tensor& self, at::optional<in
   });
 }
 
+template<typename scalar_t>
+struct NormReduction {
+  static void apply(Tensor& res, const Tensor& self, Scalar p, at::optional<int64_t> dim) {
+    auto out_ = res.data<scalar_t>();
+    auto data_ = self.data<scalar_t>();
+    auto numel = self.numel();
+    float pval = 0.0;
+    if (p.isIntegral()){
+      pval = p.to<int64_t>();
+    } else if (p.isFloatingPoint()) {
+      pval = p.to<float>();
+    }
+    if (!dim.has_value()) {
+      *out_ = reduce_all(data_, numel,  pval);
+      return;
+    }
+    int64_t n = self.size(*dim);
+    int64_t stride = self.stride(*dim);
+    // A contiguous tensor does not need to hold a meaningful stride
+    // if the corresponding size is 1
+    if (n == 1) {
+      stride = 1;
+      for (int64_t i = self.ndimension() - 1; i > *dim; i--) {
+        stride *= self.size(i);
+      }
+    }
+    int64_t batch = numel / n;
+    parallel_for(0, batch, 1, [=](int64_t begin, int64_t end) {
+      for (int64_t bi = begin; bi < end; bi++) {
+        int64_t b = bi / stride;
+        int64_t i = bi % stride;
+        const scalar_t* data = &data_[b * n * stride + i];
+        out_[bi] = norm_calc(data, n, stride, pval);
+      }
+    });
+  }
+
+  static scalar_t reduce_all(const scalar_t* data_, int64_t size,  float pval) {
+    scalar_t sum = parallel_reduce(
+      0,
+      size,
+      internal::GRAIN_SIZE,
+      (scalar_t)0,
+      [=](int64_t begin, int64_t end, scalar_t init) {
+        const scalar_t* data = &data_[begin];
+        int64_t n = end - begin;
+        scalar_t result = norm_calc(data, n, 1, pval);
+        return result;
+      },
+      std::plus<scalar_t>());
+    return sum;
+  }
+
+  static scalar_t norm_calc(const scalar_t* data, int64_t n, int64_t stride, float pval) {
+        scalar_t result = 0.0;
```

Comment:
It seems that the looks in this function can greatly benefit from the vectorized CPU instructions. 

Files under `native/cpu` are compiled multiple times, once for AVX2 enabled CPUs, once for AVX (but not AVX2) enabled CPUs, and once for CPUs with neither enabled. We provide helper class `Vec256` (see `vec256.h` and `vml.h`) to enabling writing the same code for different instruction sets. Other files under `native/cpu` (except `TensorCompareKernel`) have some examples. Maybe you can look into using that?

Reply:
@SsnL yes, some case could be vectorized e.g. p=1/2/3, while some case could not, such as p=0 or non-integer. i took a look at the vec256.h and i will update my code.

Reply:
My impression is that we don't actually need to add vectorization at this point, as this code will probably be updated once TensorIterator gets in, which will significantly simplify things.